File: 243-shortest-word-distance.java
class Solution {
    /*public int shortestDistance(String[] wordsDict, String word1, String word2) {       
        int distance = wordsDict.length;
        for(int i = 0; i < wordsDict.length; i++) {
            if(wordsDict[i].equals(word1)) {
                for(int j = 0; j < wordsDict.length; j++) {
                    if(wordsDict[j].equals(word2)) {
                        distance = Math.min(distance, Math.abs(i - j));
                    }
                }
            }
        }
        
        return distance;
    }*/

    public int shortestDistance(String[] wordsDict, String word1, String word2) {       
        int minDistance = wordsDict.length;
        int i1 = -1;
        int i2 = -1;
        
        for(int i = 0; i < wordsDict.length; i++) {
            if(wordsDict[i].equals(word1)) {
                i1 = i;
            }
            else if(wordsDict[i].equals(word2)) {
                i2 = i;
            }
            
            if(i1 != -1 && i2 != -1) {
                minDistance = Math.min(minDistance, Math.abs(i1 - i2));
            }
        }
        
        return minDistance;
    }

}

File: 243_Shortest_Word_Distance.py
class Solution:
    def shortestDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        last_position_word1 = -1
        last_position_word2 = -1
        minimum_distance = float('inf')

        for index, word in enumerate(wordsDict):
            # Update the most recent seen position of each word
            if word == word1:
                last_position_word1 = index
            elif word == word2:
                last_position_word2 = index

            # If both words have been seen, update the minimum distance
            if last_position_word1 != -1 and last_position_word2 != -1:
                minimum_distance = min(
                    minimum_distance,
                    abs(last_position_word1 - last_position_word2)
                )

        return minimum_distance
